processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,387 @@
+  {
+    if (curWriteIndex == maxSlotNum - 1) {
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextReadIndex == -1 || nextReadIndex == 0) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextWriteIndex == nextReadIndex) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+      }
+      curWriteIndex = nextWriteIndex;
+    }
+  }
+
+  @Override
+  public void reset()
+  {
+    curWriteIndex = -1;
+    nextReadIndex = -1;
+    finished = false;
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      try {
+        aggregator.close();
+      }
+      catch (Exception e) {
+        LOG.warn(e, "Could not close aggregator [%s], skipping.", aggregator);
+      }
+    }
+  }
+
+  /**
+   * Signal that no more inputs are added.  Must be called after {@link #aggregate(Object)} is called for the last input.
+   */
+  public void finish()
+  {
+    increaseWriteIndex();
+    finished = true;
+  }
+
+  /**
+   * Return a sorted iterator.  This method can be called safely while writing and iterating thread and writing thread
+   * can be different.  The result iterator always returns sorted results.  This method should be called only one time
+   * per grouper.
+   *
+   * @return a sorted iterator
+   */
+  public CloseableIterator<Entry<KeyType>> iterator()
+  {
+    if (!initialized) {
+      throw new ISE("Grouper should be initialized first");
+    }
+
+    return new CloseableIterator<Entry<KeyType>>()
+    {
+      {
+        // Waits for some data to be ready
+        final long startLoopAt = System.currentTimeMillis();
+        while ((curWriteIndex == -1 || curWriteIndex == 0) && !finished && !Thread.interrupted()) {
+          if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+            throw new RuntimeException(new TimeoutException());
+          }
+        }
+        nextReadIndex = 0;
+      }
+
+      @Override
+      public boolean hasNext()
+      {
+        return !finished || remaining() > 0;
+      }
+
+      private int remaining()
+      {
+        if (curWriteIndex >= nextReadIndex) {
+          return curWriteIndex - nextReadIndex;
+        } else {
+          return (maxSlotNum - nextReadIndex) + curWriteIndex;
+        }
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        final int recordOffset = recordSize * nextReadIndex;
+        final KeyType key = keySerde.fromByteBuffer(buffer, recordOffset);
```

Comment:
Could key and values be transferred without serialization/deserialization?

Reply:
Yes. I'm thinking to do this in a follow-up pr because it requires some refactoring of ConcurrentGrouper and SpillingGrouper which makes the patch size bigger.